### ISSUE TITLE
Crash when trying to load modded bdat files

### ISSUE THREAD

**Issue by NamelessMofo** [RoccoDev/xc3-file-loader#2]
HOS: 16.1.0
Atmosphere: 1.5.5
Game ver: 2.1.1 w/ all DLCs 1-4
XC3 file loader version: 1.3.0

I have bdat files that were dumped from bf3.ard using quickbms 0.12, and modified using bdat-toolset from this location:
https://nightly.link/RoccoDev/bdat-rs/workflows/ci/master/bdat-toolset-win64.exe.zip

When I put the files here:
sd:/atmosphere/contents/"gameid"/romfs/bdat/*.bdat

The game crashes on startup with a black screen. This happens with modified bdat files, as well as unmodified bdat files ripped directly from bf3.ard. Even 1 unmodified file in the above location causes the crash.

I will upload a copy of my atmosphere/crash_reports folder shortly.

**Comment by NamelessMofo** [comment]
Here is the crash report zip.
[crash_reports.zip](https://github.com/RoccoDev/xc3-file-loader/files/12833394/crash_reports.zip)

**Comment by RoccoDev** [comment]
Hey! Thanks for showing interest.

Could you please try these two things:

1. If you're using DLC, you should dump (and replace) BDATs from the DLC archives, not the base game. That means you should extract `bf3_dlc04.ard` instead.
2. The quickbms script caused problems for me when I originally dumped the game. Can you try getting the files using [XbTool](https://github.com/AlexCSDev/XbTool/releases/latest) instead?

**Comment by NamelessMofo** [comment]
Thanks for the super fast response! I'll give these things a try. I guess a lot has changed since the XC2/XCDE days.

**Comment by NamelessMofo** [comment]
Update:
1. I followed the advice of extracting bdats from bf3_dlc04.ard instead of the base bf3.ard.
2. As directed, I used xbtool to extract the bdats instead of quickbms. I did notice a lot cleaner result in the dump output so I'm sure quickbms was part of the problem, thank you for clearing that up.
3. I used bdat-toolset to extract sys.bdat to json.
4. I made my edits to the json file, then used bdat-toolset to repack the json's back into a new sys.bdat.
5. Copied the new bdats to SD card under romfs, preserving folder structure, ie. /atmosphere/contents/gameid/romfs/bdat/*.

The crash is now gone, but my edits are not taking effect.

**Comment by RoccoDev** [comment]
Yes, the directory layout is correct.

Apologies, I can't look into this until next week, but just to make sure this isn't caused by the edits, could you try taking an unmodified language bdat file from another language (e.g. `bdat/es/game/battle.bdat`), and renaming it (in the layeredfs directory) to `bdat/gb/game/battle.bdat`?

If everything is working and you're using English, it should replace battle translations.

For XC2, I believe xenomods has a similar feature, see "file overrides" [here] (https://github.com/BlockBuilder57/xenomods/blob/master/docs/config.md). The method I use would 100% work in DE, but it's also not required there due to lack of DLC.

**Comment by NamelessMofo** [comment]
Hi Rocco, so I tried replacing the entire `romfs/bdat/gb` folder contents with the `romfs/bdat/sp` folder contents (which includes battle.bdat plus several other bdats) and booted it up, and everything was still in English - menus, voices, battle gui, everything.

I also redownloaded the release.zip and verified that all the files from it are on my sd card, in the correct place.

I will try using the release-version-edit version of the tool and see if my title screen changes and report back.

**Comment by RoccoDev** [comment]
I tried it in Ryujinx and it seems to replace BDAT files just fine.

Could you share the directory tree of your `atmosphere/contents` directory in your SD card? (For example, open a terminal in that directory and run `tree`)

**Comment by NamelessMofo** [comment]
Sorry it took so long to reply, too much IRL. Attached the tree of my sd card atmosphere/contents/010074f013262000 folder. Of note, I do not press any override keys when I boot XC3 up. My atmosphere is set to enable mods and disable cheats by default.

Also, I tried the 1.3.1 release, thinking maybe some of my files have mismatching case, but no good.
[tree.txt](https://github.com/RoccoDev/xc3-file-loader/files/13302401/tree.txt)

**Comment by NamelessMofo** [comment]
OK I figured out the problem, and honestly it was so stupid, and I'm so angry about it, I don't even have words. I was trying to resolve an unrelated issue with some folders not showing up in another homebrew app (Checkpoint), and I remembered about the archive bit causing problems like that, so I decided to run the "Fix Archive Bit" command in Hekate (I have version 6.0.6).

Well, not only did that fix my other problem of invisible folders, but when I booted up XC3 like I always do, all my bdat mods suddenly started working. I'm extremely happy that this excellent mod now works for me, but also extremely pissed that the stupid f-ing archive bit reared its ugly head again. A serious WTF moment for me.

Anyway, *please* put this in an FAQ or the readme or something, that users should do the fix archive bit trick if they're having issues like this. I'll leave the issue open until it can be documented in the proper place.

tl;dr the archive bit can go die in a fire.

**Comment by RoccoDev** [comment]
Ah I see, I think another user solved a similar issue by reformatting the SD card.

I'll add this to the documentation, I'd have preferred to automatically fix this in the mod, but I think this comes up on the Atmosphere level and files don't make it to the layered RomFS.
